Restore action_cable.js UMD module support. Fixes #28366

javan opened the pull request:
Fix for #28366

774be3ea3b9d25ab69daf11c5071deaf053d7d5b bumped Blade from 0.6.1 to 0.7.0, which dropped [sprockets-export](https://github.com/javan/sprockets-export) from `Gemfile.lock` (Blade 0.7.0 no longer includes sprockets-export as a dependency). Doing so silently dropped `action_cable.js`'s UMD module support, essentially breaking it as an npm package.

This change restores sprockets-export and adds a new rake task that validates `action_cable.js` to help ensure this doesn't happen again.

/cc @matthewd

javan committed:
Restore action_cable.js UMD module support. Fixes #28366

matthewd commented:
:heart:

We already depend on `npm` being available in the build/release pipeline.. would it be worth extending this validation task to actually load it via nodejs somehow, and check.. something?

javan commented:
Yeah, if `node` is available we could do something like:
```diff
--- a/actioncable/Rakefile
+++ b/actioncable/Rakefile
@@ -50,6 +50,11 @@ namespace :assets do
 
     raise "[FAIL] #{file} is missing" unless pathname.exist?
     raise "[FAIL] #{file} isn't a UMD module" unless pathname.read =~ umd_pattern
+
+    require "open3"
+    stdout, stderr, status = Open3.capture3("node", "--print", "window = {}; require('#{dir}');")
+    raise "[FAIL] #{file} can't be required as a module:\n#{stderr}" unless status.success?
+
```
Which would fail like:
```
$ bundle exec rake package
Building assets…
[removed] lib/assets/compiled/action_cable.js
[created] lib/assets/compiled/action_cable.js
rake aborted!
[FAIL] lib/assets/compiled/action_cable.js can't be required as a module:
/Users/javan/Projects/rails/actioncable/lib/assets/compiled/action_cable.js:61
  ActionCable.ConnectionMonitor = (function() {
  ^

ReferenceError: ActionCable is not defined
    at Object.<anonymous> (/Users/javan/Projects/rails/actioncable/lib/assets/compiled/action_cable.js:61:3)
    at Object.<anonymous> (/Users/javan/Projects/rails/actioncable/lib/assets/compiled/action_cable.js:192:4)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at [eval]:1:14
/Users/javan/Projects/rails/actioncable/Rakefile:55:in `block (2 levels) in <top (required)>'
```
Happy to commit that if you'd like.

matthewd commented:
Sounds good :+1:

javan commented:
Updated and squashed. Output now looks like:
```
$ bundle exec rake package
Building assets…
[removed] lib/assets/compiled/action_cable.js
[created] lib/assets/compiled/action_cable.js
[verify] lib/assets/compiled/action_cable.js exists [OK]
[verify] lib/assets/compiled/action_cable.js is a UMD module [OK]
[verify] /Users/javan/Projects/rails/actioncable can be required as a module [OK]
```
